src/java.base/share/classes/jdk/internal/foreign/abi/CallBufferCache.java

```diff
@@ -26,16 +26,17 @@
 import jdk.internal.misc.Unsafe;
 import jdk.internal.vm.Continuation;
 
+import java.lang.foreign.MemorySegment;
+
 /**
- * Provides carrier-thread-local storage for up to two buffer addresses.
- * It is caller's responsibility to store homogeneous segment sizes.
- * Storing addresses, not MemorySegments turns out to be slightly faster (write barrier?).
+ * Provides carrier-thread-local storage for up to two small buffers.
  */
 public final class CallBufferCache {
     private static final Unsafe UNSAFE = Unsafe.getUnsafe();
 
     static class PerThread {
         // Two-elements to support downcall + upcall.
+        // Storing addresses, not MemorySegments turns out to be slightly faster (write barrier?).
         private long address1;
         private long address2;
```

Comment:
Can we store just `MemorySegment`s here? It would avoid having to call `ofAddress` in `acquireOrAllocate`, and would slightly reduce the footprint when using compressed oops.

Reply:
That was my original version, but this proved to be faster (albeit very little, O(.5ns)). I can't really explain why, that's above my paygrade, but one thing that comes to mind when storing references is that there's might be a GC barrier involved?

Footprint reduction would be 8 bytes * #carrier threads, and with scalar replacement, ofAddress should become a noop, right?

Let me re-run the BM once more with pointers, whether I can reproduce.

Reply:
> Footprint reduction would be 8 bytes * #carrier threads, and with scalar replacement, ofAddress should become a noop, right?

Yes.

Reply:
When moving references in and out of the cache slots, I lose .8ns:
```
Benchmark                    Mode  Cnt  Score   Error  Units
CallOverheadByValue.byPtr    avgt   10  3.283 ? 0.020  ns/op
CallOverheadByValue.byValue  avgt   10  6.309 ? 0.027  ns/op
```